torch/csrc/jit/tensorexpr/ir_simplifier.cpp

```diff
@@ -2490,6 +2490,185 @@ Stmt* SimplifierUnderContext::mutate(const For* v) {
   return new For(var_new, start_new, stop_new, body_new, loop_options);
 }
 
+// Simplify division using distributive laws for the following cases:
+// 1) (i + 24) / 6, i is the index var of a for-stmt and the range of i is
+// contained in [0, 6)
+//    -> 4
+// 2) (i + j*6) / 6, i, j are index vars of for-stmts and the range of i is
+// contained in [0, 6) + the start value of j is not less than 0
+//    -> j
+const Expr* distributeDiv(
+    const Expr* lhs,
+    const Expr* rhs,
+    VarBoundInfo var_bound_info) {
+  if (!lhs || !rhs) {
+    return nullptr;
+  }
+  auto* lhsAdd = dynamic_cast<const Add*>(lhs);
+  const Expr* rhsScalar = rhs->isConstant() ? rhs : nullptr;
+
+  if (!lhsAdd || !rhsScalar) {
+    return nullptr;
+  }
+  const Expr* lhsAdd1 = lhsAdd->lhs();
+  const Expr* lhsAdd2 = lhsAdd->rhs();
+
+  // identify index var 'i'
+  const Var* var_key = dynamic_cast<const Var*>(lhsAdd1);
+  const Expr* main = lhsAdd2;
+  if (var_key == nullptr) {
+    var_key = dynamic_cast<const Var*>(lhsAdd2);
+    main = lhsAdd1;
+  }
+
+  if (var_key != nullptr) {
+    auto got = var_bound_info.find(var_key);
+    if (got == var_bound_info.end()) {
+      return nullptr;
+    }
+
+    // check the bounds of 'i'
+    auto start = got->second.first;
+    auto end = got->second.second;
+    const Expr* check_start =
+        IRSimplifier::simplify(new CompareSelect(start, new IntImm(0), kGE));
+    const Expr* check_end =
+        IRSimplifier::simplify(new CompareSelect(end, rhsScalar, kLE));
+    if (check_start->isConstant() && check_end->isConstant() &&
+        immediateEquals(check_start, 1) && immediateEquals(check_end, 1)) {
+      const Expr* ret = IRSimplifier::simplify(new Div(main, rhsScalar));
+
+      // simplify type 1 exprs: '(i+24)/6'
+      const Expr* mod_check = IRSimplifier::simplify(new Mod(main, rhsScalar));
```

Comment:
This simplifies `(i + 24) / 6` into `4` where `i` is a non-negative int less than 6 and 24 is divisible by 6. The mod check verifies that 24 is divisible by 6. If mod check failed, we can not remove `i` in the numerator.

Reply:
Please add this as a comment as well.

Reply:
That's true. In theory, if `var%rhsScalar + main%rhsScalar < rhsScalar` then it is valid to distribute the `mod` operator: `(var + main)%rhsScalar --> var%rhsScalar + main%rhsScalar`.  In this case `(i + 25)%6 where i<5` can be simplified into `i + 1`.

But we may need to rethink the two-step approach (distribution + range based simplification) because the generalized check also requires the range info. of var and it can immediately reach the final result: `i+1` without step 2.